python/ql/src/semmle/python/dataflow/new/internal/DataFlowPublic.qll

```diff
@@ -372,6 +372,26 @@ module BarrierGuard {
         or
         this.operands(checked_node, op, str_const.getAFlowNode())
       )
+      or
+      exists(ControlFlowNode str_const_iterable, Cmpop op |
+        op = any(In in_) and safe_branch = true
+        or
+        op = any(NotIn ni) and safe_branch = false
+      |
+        this.operands(checked_node, op, str_const_iterable) and
+        (
+          str_const_iterable instanceof SequenceNode
+          or
+          str_const_iterable instanceof SetNode
+        ) and
+        forall(ControlFlowNode elem |
+          elem = str_const_iterable.(SequenceNode).getAnElement()
+          or
+          elem = str_const_iterable.(SetNode).getAnElement()
+        |
+          elem.getNode() instanceof StrConst
```

Comment:
This bit bugs me a bit, because it seems strangely inverted. Would it perhaps improve matters to define a separate `abstract class` of "constant iterables containing strings" and then extending it with subclasses for sequences and sets? 

In particular, one thing that's unfortunate is that this doesn't account for local flow, which might be useful. One _could_ extend the above to take local flow into consideration, but it would really blow up the size (in term of lines of code) of the characteristic predicate.

This also makes me think that perhaps these things should not live in `DataFlowPublic` (which is getting a bit crowded), but perhaps in a file dedicated solely to barrier guards?

Reply:
I agree. that code did not turn out that nice. I've tried to rewrite, let me know what you think :)